### ISSUE TITLE
Calculate moon ecliptic longitude 

### ISSUE THREAD

**Issue by shawnlauzon** [ryanseys/lune#29]
Hello, you've got a great project here in that it tells me when the moon will enter a new phase; really awesome, exactly what I need. I'm also looking for the same type of functionality, but for the longitude of the ecliptic (which I would use for calculating the zodiac sign that the moon will be in at that point). Any thoughts on how this could be added? I would be happy to write the software to do this, but I have zero knowledge of the math which would be behind it. Thank you 🙏

**Comment by ironwallaby** [comment]
It would be another couple equations (along the lines of https://github.com/ryanseys/lune/blob/master/lib/lune.js#L96-L103) to calculate the moon's "mean longitude" and "equation of the center" (which, when added together, give the moon's ecliptic longitude).

This isn't hard, but it is complicated: the moon's close distance and low mass means that a lot of periodic terms (to account for the pull of the sun, venus, mars, etc.) are needed to get reasonable accuracy.

Furthermore, if you're interested in the moon's _apparent_ ecliptic longitude (rather than the geometric ecliptic longitude), you will need to take into account _parallax_, since the moon's so close (relative to the size of the earth) that where you're located on the earth makes a big difference!

The best source for this that I'm aware of is Astronomical Algorithms, 2nd ed., which is easy to find (e.g. on Amazon) and is worth study.

Good introductory sources (though I might not use them for this library, since they use different and/or incomplete source data which might cause accuracy issues) are [Wikipedia](https://en.wikipedia.org/wiki/Lunar_theory) and [Paul Schlyter's website](http://stjarnhimlen.se/comp/ppcomp.html#9).

I have a different library that computes this [`astro`](https://github.com/darkskyapp/astro) if you're interested, though it may be overkill for your use case!

**Comment by shawnlauzon** [comment]
Thanks @ironwallaby for the fast response. I was able to find some implementations of Astronomical Algorithms, but you're right I haven't yet read the original source to understand what is going on. 

It seems like the complicated part here is to determine when the moon enters a certain sign in the future (like this library does for phases of the moon). For example [this table](https://www.moontracks.com/lunar_ingress.html) lists the dates and times for each of the signs, but for obvious reasons I don't want to hard-code all those values into the library.

I actually tried to use `astro` for this, but I get negative or very large numbers; the api says `ms`, I believe that means that this should work:

```javascript
console.log(astro.lunar.longitude(new Date().valueOf()));
```

**Comment by ironwallaby** [comment]
Yes, the lunar longitude increases without bound: you will need to modulus it into the 0..360° range!

```javascript
// get longitude
let longitude = astro.lunar.longitude(Date.now());

// convert from radians to degrees
longitude = longitude * (180 / Math.PI);

// wrap from 0..360
// (this is like "longitude % 360" but always returns a positive number)
longitude = longitude - Math.floor(longitude / 360) * 360;
```

The simplest way to determine the time when the moon has a particular longitude would be to do (e.g.) a bisection search; a more sophisticated way would be to invert the formula (e.g. solve the formula for longitude) and then do something like Newton-Raphson. (This will be faster, but adds a lot of complexity and probably isn't worth it unless you _really_ want to learn the math :) )

**Comment by shawnlauzon** [comment]
Ahh, thank you! I'll check that out in the morning. As for doing a search, that would be interesting ... to get down to seconds precision. Would need to brush up on either math or algorithms seems like. You've been a big help and I'll check back here after I have some progress. 🙏

**Comment by shawnlauzon** [comment]
Thanks @ironwallaby that code works great. Unfortunately, the precision is less than what I need; I need the 10 second precision provided by AA. The only place I have found something which is giving me accurate results is [this page from 1997](http://www.shallowsky.com/moon/). The code is licensed GPL and so eventually I'll have to reimplement it.

As for Newton-Raphson, all that I know is what I gleaned from wikipedia, and it's far beyond my memory of math from college days. Since I know each sign is about 2.5 days later, shouldn't be too bad to find them.
